### PR TITLE
API: Product Reviews #55

### PR DESCRIPTION
**Pull request by sky-hub**
### Description (*)
Service contracts for product reviews.

As discussed with @paliarush, defined services are implemented similar to MSI implementation and allow bulk operations.
The difference in this implementation is that the service is returning a Response object containing success,  failed, retryable object and errors associated to failed or retryable objects. Currently retyrable objects are not returned as I could not identify a specific error for this case (a database deadlock may be a case but furher research is needed)

1. Added validation chain for review validate
1. Added review aggregation service: app/code/Magento/Review/Model/Aggregator.php
1. Updated method names to make them more specific and forwarded magic methods to new methods
Ex: magic method $review->getStatusId() was forwarded to $review->getStatus()
1. Some methods are not defined using strict types in order to keep backwards compatibility
1. Added updated_at column for Review model
1. Updated review tests
1. Added support for extension attributes for reviews and related entities
1. Added the possibility to save review ratings based on rating name and rating value. This was implemented to allow review ratings to be added without knowing the internal ids and also to align code base with the GraphQl implementation.

Notes:
1. ReviewGraphQl module needs update after this PR is merged
1. CreateReviews service can not be used to insert multiple reviews at the same time because review ratings need the reviewId when saving.
1. Some methods can not enforce strict types as they need to remain backwards compatible

### Fixed Issues (if relevant)
1. https://github.com/magento/community-features/issues/55 - API: Product Reviews
2. https://github.com/magento/magento2/issues/9570 - Product Review API

### Manual testing scenarios (*)
1. CREATE Reviews: make a post request to https://store.domain/rest/V1/create-reviews with the following data:
```
 {"reviews":[{"review_id":1,"related_entity_id":1,"review_entity_id":1,"customer_nickname":"Nickname","title":"Review Summary","review_text":"Review text","ratings":[{"rating_name":"Quality","rating_value":2},{"rating_name":"Price","rating_value":3}],"store_id":1,"stores":[1],"created_at":"0000-00-00 00:00:00","updated_at":"0000-00-00 00:00:00"},{"review_id":2,"related_entity_id":1,"review_entity_id":1,"customer_nickname":"Nickname","title":"2 filter first review","review_text":"Review text","ratings":[{"rating_name":"Price","rating_value":3}],"store_id":1,"stores":[1],"created_at":"0000-00-00 00:00:00","updated_at":"0000-00-00 00:00:00"}]}
```

2. CREATE Reviews: make a post request to https://store.domain/rest/V1/create-reviews with the following data:
```
{"reviews":[{"review_id":1,"related_entity_id":1,"review_entity_id":1,"customer_nickname":"Updated Nickname","title":"Updated Review Summary","review_text":"Updated Review Text","ratings":[{"rating_name":"Quality","rating_value":2},{"rating_name":"Price","rating_value":3}],"store_id":1,"stores":[1],"created_at":"0000-00-00 00:00:00","updated_at":"0000-00-00 00:00:00"},{"review_id":2,"related_entity_id":1,"review_entity_id":1,"customer_nickname":"Updated Customer Nickname","title":"Updated Review Title","review_text":"Updated Review Comment","ratings":[{"rating_name":"Price","rating_value":3}],"store_id":1,"stores":[1],"created_at":"0000-00-00 00:00:00","updated_at":"0000-00-00 00:00:00"}]}
```

3. GET Reviews: make a POST request to https://store.domain/rest/V1/get-reviews?searchCriteria[filterGroups][0][filters][0][field]=sku&searchCriteria[filterGroups][0][filters][0][value]='SOME-SKU'

4. DELETE Reviews: make a POST request to https://store.domain/rest/V1/delete-reviews with the following data:
```
{"reviews":[{"review_id":1,"related_entity_id":1,"customer_nickname":"Nickname","title":"Review Summary","review_text":"Review text","created_at":"0000-00-00 00:00:00","updated_at":"0000-00-00 00:00:00"},{"review_id":2,"related_entity_id":1,"customer_nickname":"Nickname","title":"2 filter first review","review_text":"Review text","created_at":"0000-00-00 00:00:00","updated_at":"0000-00-00 00:00:00"}]}
```

5. Related Links:
 - Review GraphQL implementation: https://github.com/magento/graphql-ce/pull/227

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds on Travis CI are green)

### CODE REVIEW

**Comment by magento-cicd2** [comment]
[CLA assistant check](https://cla.dev.magento.com/magento/magento2?pullRequest=19266) <br/>All committers have signed the CLA.

**Comment by magento-engcom-team** [comment]
Hi @sky-hub. Thank you for your contribution 
Here is some useful tips how you can test your changes using Magento test environment.
Add the comment under your pull request to deploy test or vanilla Magento instance:
- `@magento-engcom-team give me test instance` - deploy test instance based on PR changes
- `@magento-engcom-team give me 2.3-develop instance` - deploy vanilla Magento instance

For more details, please, review the [Magento Contributor Assistant](https://devdocs.magento.com/guides/v2.2/contributor-guide/contributing.html#contributor-assist) documentation

**Comment by okorshenko** [comment]
Hi @sky-hub 
Could you please sign Contributor License Agreement  so that we can proceed with PR review
https://cla.dev.magento.com/magento/magento2?pullRequest=19266

**Comment by sky-hub** [comment]
> 
> 
> Hi @sky-hub
> Could you please sign Contributor License Agreement so that we can proceed with PR review
> https://cla.dev.magento.com/magento/magento2?pullRequest=19266

Hello @okobchenko, sorry, fortgot about it. I would like to point that this PR is just a preview at what needs to be done on the long run. There is also the option to define APIs and implement them but I wouldn't take that approach since the current implementation is far from what it should be and most probably it would complicate things on long term if we keep postpone refactoring it.

I have added some TODOs where I considered that code needs to be updated.

File: app/code/Magento/Review/Api/Data/ReviewInterface.php
Comment (paliarush) [inline app/code/Magento/Review/Api/Data/ReviewInterface.php]: Since `Ratings` functionality is part of `Review` module there is no need to define extension attributes. In general extension attributes can help when functionality of one module needs to be extended with the functionality of the other module.

File: app/code/Magento/Review/Api/Data/ReviewInterface.php
Comment (paliarush) [inline app/code/Magento/Review/Api/Data/ReviewInterface.php]: Current naming is confusing and should be improved when we introduce new API.

`entityId` is more like `relatedEntityTypeId` 

`entityPkValue` is actually `relatedEntityId` 

Please see if you can come up with better naming.

File: app/code/Magento/Review/Api/ReviewRepositoryInterface.php
Comment (paliarush) [inline app/code/Magento/Review/Api/ReviewRepositoryInterface.php]: According to technical vision, all new services must support bulk operations (including save, update and delete). 
As a result we should have 4 interfaces for get, create, update and delete.

File: app/code/Magento/Review/Model/Review.php
Comment (paliarush) [inline app/code/Magento/Review/Model/Review.php]: It affects caching, we don't want to make it configurable.

File: app/code/Magento/Review/Api/ReviewRepositoryInterface.php
Comment (paliarush) [inline app/code/Magento/Review/Api/ReviewRepositoryInterface.php]: Please split this repository into individual interfaces like it is done in MSI https://github.com/magento-engcom/msi/tree/2.3-develop/app/code/Magento/InventoryApi/Api

Skip repository for now.

Please pay attention to the return types being `void` to enable asynchronous invocation in the future if necessary.

File: app/code/Magento/Review/Model/Review.php
Comment (paliarush) [inline app/code/Magento/Review/Model/Review.php:36]: Lets try to avoid any changes in the existing models except for `AbstractModel` => `AbstractExtensibleModel` change and new interface.

File: app/code/Magento/Review/Api/ReviewRepositoryInterface.php
Comment (paliarush) [inline app/code/Magento/Review/Api/ReviewRepositoryInterface.php]: Please split `save` into `create` and `update` because they have different validation rules. Common logic should be reused.

**Comment by sivaschenko** [comment]
Hi @sky-hub are you still working on this pull request? Please let us know when it's ready for review.

**Comment by nuzil** [comment]
Hi @sivaschenko , @sky-hub needed some support to define main direction of current implementation. We had several talks with @paliarush and I guess now @sky-hub could proceed with this PR 👍

**Comment by sky-hub** [comment]
Due to holidays, etc, this PR is going to be put on hold until 15th of January 2019. I will continue working after that date.

File: app/code/Magento/Review/Model/Command/CreateReviews.php
Comment (sivaschenko) [inline app/code/Magento/Review/Model/Command/CreateReviews.php:93]: I'd suggest to accumulate valid and invalid review using corresponding variables in the foreach loop and perform create operation or add failed review to response in bulk after the foreach.

File: app/code/Magento/Review/Model/Command/GetReviews.php
Comment (sivaschenko) [inline app/code/Magento/Review/Model/Command/GetReviews.php]: Please change the access level of these properties to private

File: app/code/Magento/Review/Model/Command/GetReviews.php
Comment (sivaschenko) [inline app/code/Magento/Review/Model/Command/GetReviews.php]: Please consider using virtual type of `Magento\Framework\Api\SearchCriteria\CollectionProcessor\FilterProcessor` for filter processing

File: app/code/Magento/Review/Model/Command/UpdateReviews.php
Comment (sivaschenko) [inline app/code/Magento/Review/Model/Command/UpdateReviews.php]:
```suggestion
                unset($reviews[$key]);
```

Comment (sivaschenko) [reply app/code/Magento/Review/Model/Command/UpdateReviews.php]: However, in my opinion, it's better to use a separate variable for accumulating valid/invalid reviews

File: app/code/Magento/Review/Model/Rating/Option.php
Comment (sivaschenko) [inline app/code/Magento/Review/Model/Rating/Option.php:48]: `setId` method should be preserved for backward compatibility

File: app/code/Magento/Review/Model/Rating/Option/Vote.php
Comment (sivaschenko) [inline app/code/Magento/Review/Model/Rating/Option/Vote.php:11]: I am not sure if it is appropriate to remove `@api` annotation from classes

File: app/code/Magento/Review/Model/ResourceModel/Review/CreateMultiple.php
Comment (sivaschenko) [inline app/code/Magento/Review/Model/ResourceModel/Review/CreateMultiple.php]: In my opinion, it will be better to move the try-catch inside the foreach and accumulate errors for all the failed reviews create attempts

File: app/code/Magento/Review/Model/ResourceModel/Review/CreateMultiple.php
Comment (sivaschenko) [inline app/code/Magento/Review/Model/ResourceModel/Review/CreateMultiple.php]: I find it confusing to use such hardcoded variables as `1`, in my opinion, it's better to use `1` as a default parameter value for functions designed to be called with such arguments

File: app/code/Magento/Review/Model/ResourceModel/Review/CreateMultiple.php
Comment (sivaschenko) [inline app/code/Magento/Review/Model/ResourceModel/Review/CreateMultiple.php]: Please consider reusing framework select builder (i.e. `\Magento\Framework\DB\Adapter\AdapterInterface::insert`)

**Review by sivaschenko — COMMENTED** [review]
Hi @sky-hub thanks for the contribution! I've done a partial brief review at this point, however, as I can see there is still work to be done on this pull request. Please see my code review comments

**Comment by sidolov** [comment]
Hi @sky-hub , are you still working on the pull request?

**Comment by sky-hub** [comment]
> 
> 
> Hi @sky-hub , are you still working on the pull request?

Yes. Starting from today I will continue working on this pull request.

File: app/code/Magento/Review/Model/Rating/Option.php
Comment (sky-hub) [reply app/code/Magento/Review/Model/Rating/Option.php:48]: setId is already implemented in parent classes: lib/internal/Magento/Framework/Model/AbstractExtensibleModel.php and lib/internal/Magento/Framework/Model/AbstractModel.php, that is why I removed it.

File: app/code/Magento/Review/Model/Rating/Option/Vote.php
Comment (sky-hub) [reply app/code/Magento/Review/Model/Rating/Option/Vote.php:11]: I had it removed from model and added it to interface, I think I had a discussion about this with @paliarush, but I will double check just to be sure.

File: app/code/Magento/Review/Model/Command/CreateReviews.php
Comment (sky-hub) [reply app/code/Magento/Review/Model/Command/CreateReviews.php:93]: Ok, it can be done.

File: app/code/Magento/Review/Model/Rating/Option/Vote.php
Comment (sky-hub) [reply app/code/Magento/Review/Model/Rating/Option/Vote.php:11]: It is ok to be removed and added to interface.

**Comment by PiotrSiejczuk** [comment]
Hi @sky-hub are you still planning to continue the great work you are doing on this topic? I am currently facing a requirement @ one project to have such Reviews API in place. Would be good to know if you are planning to finish this PR :) Thank you in advance for your replay!

**Comment by sky-hub** [comment]
> 
> 
> Hi @sky-hub are you still planning to continue the great work you are doing on this topic? I am currently facing a requirement @ one project to have such Reviews API in place. Would be good to know if you are planning to finish this PR :) Thank you in advance for your replay!

Hello @PiotrSiejczuk, job's done. This PR can entered the code review process starting from now.

PS: @paliarush Please let me know when this PR is merged in the main line, I have to make some updates for the GraphQl implementation also.

Cheers

**Review by sivaschenko — CHANGES_REQUESTED** [review]
Hi @sky-hub if the pull request is finally completed - please resolve conflicts and squash all the commits into a single one

**Comment by sky-hub** [comment]
> 
> 
> Hi @sky-hub if the pull request is finally completed - please resolve conflicts and squash all the commits into a single one

Conflict solved.

**Comment by sivaschenko** [comment]
I'd suggest that probably the better location for the tests that are currently in `Magento\ReviewApi\Test\Api` is `Magento_Review` module as they are actually testing the implementation of the API.

@akaplya can you please provide your opinion regarding the above as Review appears to be your component

File: app/code/Magento/Review/Model/Review/RatingsProcessor.php
Comment (afirlejczyk) [inline app/code/Magento/Review/Model/Review/RatingsProcessor.php]: We should probably load rating title (name) for specific store ?
If so, we should add
`  $collection->addRatingPerStoreName($storeId);`

**Comment by lenaorobei** [comment]
Unfortunately, I need to close this PR, since "Product Reviews" functionality was added in PR https://github.com/magento/magento2/pull/27882 based on architectural design from https://github.com/magento/architecture/pull/298.

**Pull request merged**
